wechatpay/src/main/java/com/stripe/android/payments/wechatpay/WeChatPayAuthenticator.kt

```diff
@@ -0,0 +1,65 @@
+package com.stripe.android.payments.wechatpay
+
+import androidx.activity.result.ActivityResultCallback
+import androidx.activity.result.ActivityResultCaller
+import androidx.activity.result.ActivityResultLauncher
+import com.stripe.android.model.StripeIntent
+import com.stripe.android.model.getRequestCode
+import com.stripe.android.networking.ApiRequest
+import com.stripe.android.payments.PaymentFlowResult
+import com.stripe.android.payments.core.authentication.IntentAuthenticator
+import com.stripe.android.view.AuthActivityStarterHost
+
+/**
+ * [IntentAuthenticator] authenticating through WeChatPay SDK.
+ */
+class WeChatPayAuthenticator : IntentAuthenticator {
+    /**
+     * [weChatPayAuthLauncher] is mutable and might be updated during
+     * through [onNewActivityResultCaller]
+     */
+    private var weChatPayAuthLauncher: ActivityResultLauncher<WeChatPayAuthContract.Args>? = null
+    private val weChatAuthLauncherFactory =
+        { host: AuthActivityStarterHost, requestCode: Int ->
+            weChatPayAuthLauncher?.let {
+                WeChatPayAuthStarter.Modern(it)
+            } ?: WeChatPayAuthStarter.Legacy(host, requestCode)
+        }
+
+    override fun onNewActivityResultCaller(
+        activityResultCaller: ActivityResultCaller,
+        activityResultCallback: ActivityResultCallback<PaymentFlowResult.Unvalidated>
+    ) {
+        weChatPayAuthLauncher = activityResultCaller.registerForActivityResult(
+            WeChatPayAuthContract(),
+            activityResultCallback
+        )
+    }
+
+    override fun onLauncherInvalidated() {
+        weChatPayAuthLauncher?.unregister()
+        weChatPayAuthLauncher = null
+    }
+
+    override suspend fun authenticate(
+        host: AuthActivityStarterHost,
+        stripeIntent: StripeIntent,
+        threeDs1ReturnUrl: String?,
+        requestOptions: ApiRequest.Options
+    ) {
+        val weChatPayRedirect =
+            requireNotNull(
+                stripeIntent.nextActionData as? StripeIntent.NextActionData.WeChatPayRedirect
+            ) {
+                "stripeIntent.nextActionData should be WeChatPayRedirect, instead it is " +
+                    "${stripeIntent.nextActionData}"
+            }
+
+        weChatAuthLauncherFactory(
+            host,
+            stripeIntent.getRequestCode()
```

Comment:
The `requestCode` feels like something that should be passed via `authenticate()` - is that possible?

Reply:
hmm the `StripeIntent` is also used to extract nextActionData here, I think passing the entire object is more appropriate here.
I do plan to refactor the #authenticate a bit to support `SetupIntent` and [Source](https://github.com/stripe/stripe-android/blob/916a6516fc69658de7d3740959033fba6e009549/payments-core/src/main/java/com/stripe/android/StripePaymentController.kt#L350), will possiblely clean up the `authenticate` param list in that PR